### ISSUE TITLE
[Question] What are the advantages of using numba instead of native pytorch while computing RNNT-Loss?

### ISSUE THREAD

**Issue by manneh** [NVIDIA/NeMo#3260]
Hi . I saw that the pretrained Transducer model uses Numba implementation of RNNT-Loss. I was wondering what is the reason behind using numba instead of implementing it in native pytorch?

**Comment by titu1994** [comment]
RNNT loss is very slow to compute using native pytorch, especiay on the GPU since it involves 4th order loop - batch x acoustic length x text length x vocabulary. Cuda can parallelize over all the dimensions at once efficiently. 

Further the gradient of the explicit loop based pytorch loss would take an enormous amount of time as well, which would slow down training. So we use Numba + CUDA to accelerate the loss and gradient calculation. 

Finally, computing the explicit gradient during the forward call itself (without having to call .backward()) allows batch splitting allowing for larger batch sizes to be used during training v
